core/trino-main/src/main/java/io/trino/operator/output/PagePartitioner.java

```diff
@@ -147,7 +154,26 @@ public void partitionPage(Page page)
             return;
         }
 
-        if (page.getPositionCount() < partitionFunction.getPartitionCount() * COLUMNAR_STRATEGY_COEFFICIENT) {
+        boolean partitionByRow;
```

Comment:
nit: you could extract this logic to a separate method. `partitionPage` has this simple structure that delegates to either row or columnar strategy and the `partitionByRow` calculation logic breaks this.